### PR TITLE
test(search): tighten product timeout assertion

### LINKED ISSUE #530: Tighten enumerative product timeout-test assertion to match the length-two sibling
Deferred from #489: tighten the assertion in `run_length_product_sets_stop_when_nested_deadline_expires`.

Original feedback by @claude (automated review) on PR #489 (https://github.com/pmatos/s11/pull/489):

> With 2 instructions and length 3, the full sweep is 2^3 = 8 candidates, so `< 8` only asserts "stopped before the very last candidate". Given that `sequence_cost` sleeps 50ms and the timeout is 25ms, in practice this should stop after exactly 1 cost call. The analogous `run_length_two` test uses `assert_eq!(..., 1, ...)`. Tightening to `< 2` (or `== 1`) would give a more precise failure message.

**Context:** `src/search/enumerative/search.rs` — `run_length_product_sets_stop_when_nested_deadline_expires` asserts `cost_calls < 8`; the sibling `run_length_two_sets_stop_when_inner_deadline_expires` asserts `== 1`. The trace is deterministic (`cost_calls == 1`) with the same 50ms-vs-25ms 2x margin, so `== 1` would be reliable.

**Why deferred:** diminishing-returns — a test-only consistency tweak that the *first* automated review of this same PR explicitly called "timing-sensitive ... Pre-existing pattern, not a new problem." Not worth re-triggering a full CI/review round on an already-green PR; fold into a future test-cleanup pass.

_Filed automatically by `pm-autofix-pr` after dual-evaluator triage by Claude Opus 4.8 and Codex GPT-5.5._

### PR DESCRIPTION
**Pull request by pmatos**
## Summary

- tighten the nested length-product timeout test to reject evaluating a second candidate
- allow zero cost calls when the deadline expires during Rayon dispatch or recursive polling
- preserve the separate assertion that the nested timeout sets the shared stop flag

## Testing

- targeted product and sibling timeout tests under the default harness
- both timeout tests under `cargo llvm-cov`
- `cargo fmt --all`
- `cargo clippy --all -- -D warnings`
- `cargo test --all`
- `./ci_check.sh`

Closes #530

**Merge with `gh pr merge --merge` (merge commit, not squash) per CLAUDE.md.**

### CODE REVIEW

**Comment by pmatos** [comment]
Automation-contract note: `scripts/full_test.sh` is Vow-specific and does not exist in this s11 repository, so I used the repository-authoritative `./ci_check.sh`, which passed its full suite. The plan-only `cargo clippy --all-features --all-targets -- -D warnings` variant also reports two pre-existing `clone_on_copy` warnings at `src/x86_window_reassembly.rs:245` and `:280`; both lines are unchanged from `origin/main`. I left them untouched to preserve issue #530’s one-file test-only scope. The contract-required `cargo clippy --all -- -D warnings` passes.

**Comment by pmatos** [comment]
## Automated code-review pass

Ran `/code-review --fix` against this PR. **Verdict on the change itself: correct.**

I verified the new `assert_eq!(cost_calls, 1)` empirically rather than by inference from the length-two sibling, since the two have different loop structures. Trace: `all_instructions` has 2 entries, `length: 3`, and `SharedState::new(0)` gives `best_cost == 0`, so the fake `sequence_cost` return of `1` prunes before `check_equivalence` (which would `panic!`). The first `sequence_cost` sleeps 50ms, blowing the 25ms budget, so the next `stop_if_timed_out()` sets `stop` and every recursion level unwinds. Exactly 1 cost call — not 0, not 2.

### Fix applied (353ebd6)

The new message `"timeout should stop after exactly 1 cost call; saw {cost_calls}"` duplicated what `assert_eq!` already prints as `left`/`right`, while dropping the rationale the old `< 8` message carried. The sibling assertion explains *why* (`"timeout should be checked before evaluating the second inner candidate"`); true parity means an explanatory message, not a redundant one. Rewrote it to state the 50ms-vs-25ms reason and keep the `8 = 2^3` full-sweep baseline.

### Flake check

Worth stating the asymmetry explicitly: `cost_calls < 8` passed when the count was 0, but `assert_eq!(cost_calls, 1)` fails when the count is 0. The new assertion is therefore strictly more flake-prone in the zero-call direction, and the zero-call path is a >25ms stall before the first `stop_if_timed_out()` — i.e. CPU contention. So filtered serial runs are the *least* contended condition and not the one that would trigger the failure. Runs actually performed on this workspace, all green:

- 15x both timeout tests, filtered, `--test-threads=1` — 0 failures
- 5x full `cargo test --lib` at default parallelism, 1692 tests each (the real CI condition) — 0 failures
- 40x both timeout tests under 72 busy-loop threads on 24 cores (3x oversubscription) — 0 failures

`./ci_check.sh` also passes in full against the post-fix tree.

### Findings deliberately not acted on

**Arity is not pinned.** `cost_calls == 1` does not observe the *length* of the sequence handed to `sequence_cost`. If `enumerate_suffix`'s `candidate.len() == self.length` guard regressed to fire at depth 1, the count would still be 1 and the test would still pass while the product search silently stopped enumerating length-three candidates. Fixing this means adding an `AtomicUsize` to `InnerTimeoutIsa` — shared infrastructure both timeout tests depend on — for a concern #530 did not raise. Judged out of scope for a 3-line consistency tweak; worth a follow-up issue if the arity guard is considered at risk.

**Timing sensitivity left as-is.** Per the runs above it is not reproducible here, and #530 explicitly triaged this as a pre-existing pattern rather than a new problem. The only "fix" would revert the change this PR exists to make, so it stands as intended.

**Pre-existing, outside the hunk** (not fixed, noted for the record): `reset_inner_timeout_counter`'s `.expect("inner timeout test lock poisoned")` turns a failure in one timeout test into a misleading poison panic in the other; and `run_length_product`'s `Vec::with_capacity(length + terminator)` is wasted because `evaluate_candidate` clones the candidate (capacity == len) before pushing the terminator.

**Comment by pmatos** [comment]
## Simplify pass

Ran `/simplify` over this PR's diff across four angles (reuse, simplification, efficiency, altitude). One finding applied, one skipped.

**Applied** — `3210089`: inlined the vestigial `cost_calls` binding in `run_length_product_sets_stop_when_nested_deadline_expires`. It existed to feed the `saw {cost_calls}` interpolation that `353ebd6` removed; `assert_eq!` prints its left operand itself, and `pool.install` has already joined before the load, so the snapshot bought nothing. Inlining makes the assertion site read identically to the length-two sibling, which is the point of this PR. Flagged independently by both the simplification and altitude angles.

**Skipped** — the altitude angle proposed trimming the assertion message to one clause and relocating the 50ms/25ms mechanism into a comment on `InnerTimeoutIsa::sequence_cost`. Declined for two reasons: the fix edits `sequence_cost` (~90 lines outside the reviewed diff), and it would delete the entire payload of `353ebd6`, whose stated purpose was to add that explanation. Its underlying observation is fair — the message asserts as fact something untrue when the assertion fires — but `assert_eq!` prints the actual value directly above the message, so a failing reader sees `left: N, right: 1` alongside the rationale for expecting 1. Recorded here rather than actioned so a future pass doesn't rediscover it.

**Clean** — reuse and efficiency found nothing.

Verified the two facts the assertion message states: `sequence_cost` sleeps 50ms against the test's 25ms budget, and `enumerate_all` returns 2 instructions, so length three is 2³ = 8 candidates. Both check out.

`./ci_check.sh` green; both sibling tests pass.

File: src/search/enumerative/search.rs [hunk 1]
```diff
@@ -1735,9 +1735,9 @@ mod tests {
             shared.stop.load(Ordering::Relaxed),
             "nested product loop should set stop after the timeout expires"
         );
-        assert!(
-            cost_calls < 8,
-            "timeout should stop before the full length-three product sweep; saw {cost_calls} cost calls"
+        assert_eq!(
+            cost_calls, 1,
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/search/enumerative/search.rs]: Addressed in 547edce. The product-timeout oracle now keeps the separate `shared.stop` assertion and requires `cost_calls < 2`, so it still rejects reaching a second product candidate while accepting the valid zero-call schedule when the 25ms deadline expires during Rayon dispatch or recursive pre-leaf polling. I considered synchronizing entry into the first cost call or adding a controllable clock, but that would add a production seam solely to test a minimum-work guarantee the optimizer does not make. Both timeout tests pass normally and under `cargo llvm-cov`; `cargo fmt --all`, `cargo clippy --all -- -D warnings`, `cargo test --all`, and `./ci_check.sh` are green. The Vow-specific `scripts/full_test.sh` named by the generic run contract is absent from s11, so the repository-authoritative `./ci_check.sh` was used for the full local gate.

**Comment by pmatos** [comment]
Confirmed intentional. The zero-call scheduling argument can apply to the untouched length-two sibling as well; that pre-existing test is already tracked separately in #454. This PR limits itself to #530 and the product assertion it changed: `< 2` is one of #530’s proposed precise bounds, preserves the new “never evaluate a second candidate” invariant, and fixes the zero-call regression introduced by this PR’s earlier `== 1` edit. Broadening the diff to the sibling would combine two independently tracked test-oracle cleanups.

**Comment by pmatos** [comment]
## Automated code-review pass — `/code-review --fix` (xhigh effort)

Ran against the PR diff (`origin/main...HEAD`, 8 lines, test-only, `src/search/enumerative/search.rs`). **5 findings surfaced, 0 fixes applied.** Working tree clean, `cargo fmt --check` clean, both timeout tests pass. Recording the judgment call so a reviewer can override.

### Findings, and why each was skipped

1. **Assertion asymmetry with the length-two sibling** (line ~1692). The product test now allows `cost_calls < 2` while `run_length_two_sets_stop_when_inner_deadline_expires` still asserts `== 1`. The review verified that `run_length_two` has the *same* pre-first-candidate window (`par_iter` dispatch + two stop checks before `evaluate_candidate`), so the zero-call rationale in the new comment applies to both — the review asked us to "pick one story".
   **Skipped:** issue #530's own body accepts either bound verbatim — *"Tightening to `< 2` (or `== 1`) would give a more precise failure message."* Moving to `== 1` would revert commits `547edce` ("allow timeout before first product candidate") and `e72a477` ("document zero-call timeout schedule"), which settled this deliberately; loosening the sibling would weaken a neighbouring assertion the reviewer measured as stable (1 cost call in 27/27 runs, 12 of them under 24-way CPU contention). Neither is a call to make inside a review-application pass.

2. **No lower bound: the test passes vacuously at `cost_calls == 0`** (line ~1740). A regression making `ProductContext::stop_if_timed_out` always return true, or `enumerate_suffix` never reach `candidate.len() == self.length`, yields `stop=true, cost_calls=0` and both assertions still pass.
   **Skipped:** adding `cost_calls > 0` directly contradicts this PR's stated intent. The non-contradictory alternative (sleep-toggle flag plus a new no-timeout product test pinning all 8 candidates) is a fixture redesign disproportionate to an 8-line test diff. Worth a follow-up issue if a reviewer disagrees.

3. **Pre-existing — `INNER_TIMEOUT_TEST_LOCK` poisoning masks real failures** (line ~1622). `reset_inner_timeout_counter` uses `.expect("inner timeout test lock poisoned")`, so a panic in either timeout test makes the sibling report a poison error instead of its own assertion diff. `CACHE_PROBE_TEST_LOCK` (~900) and `VERIFY_STATS_TEST_LOCK` (~1344) share the pattern, so a fix (`unwrap_or_else(|e| e.into_inner())`) should be file-wide.
   **Skipped:** pre-existing, untouched by this diff, and a correct fix reaches well outside it.

4. **Pre-existing — ~25 lines of duplicated setup** between the two timeout tests (~1662-1685 vs ~1701-1731); only the runner call and target length differ. This duplication is what let the two copies drift in the first place.
   **Skipped:** pre-existing refactor, outside the diff.

5. **Pre-existing — dead terminator reservation in `run_length_product`** (~649). The `length + terminator.is_some()` reserve is discarded by the `candidate.clone()` at ~616 (`Vec::clone` allocates `capacity == len`), so every leaf reallocates when a terminator is pinned.
   **Skipped:** production-code change in the hot enumeration path, unrelated to a test-assertion PR.

### Also checked and clean

`\`-continuation in the new assert message renders with a single space; `cargo fmt --check` passes; no stale `< 8` references anywhere in repo or docs; no co-author attribution in commit messages; no `PLAN.md` staged.

_Posted by an unattended review run. Findings 1-2 are author calls — override freely._

**Comment by pmatos** [comment]
Automated `/simplify` pass (quality only, not a bug hunt). Four review angles ran: **reuse**, **efficiency**, and **altitude** returned no findings; **simplification** returned two, both applied in af36b25.

1. **Comment placement.** The zero-call rationale added in e72a477 sat above the `cost_calls` load but explained the `< 2` bound on the `assert!` two lines below. Moved it directly above the `assert!` it justifies.

2. **Vestigial line continuation.** The `\`-wrapped failure message was left over from the earlier three-line message that 547edce shortened. Unwrapped to a single line.

**Judgment call worth recording:** the simplification and altitude agents disagreed on item 2. The altitude agent defended the wrap as matching precedent at `search.rs:2844`. I overrode that: 2844 unwrapped would be 136 columns, so its wrap is *forced*, whereas this literal is 110 and is not. The decisive evidence is that `origin/main`s version of this very assert was already a single 106-column line, and `search.rs:1948` is another 106-column single-line assert message that passes CI — so removing the wrap restores the files existing style rather than introducing a new one. It also removes the trailing-space-before-backslash hazard, where a future edit could silently render `saw1 cost calls`. The alternative considered was shortening the message to fit under 100 columns; rejected because it would have dropped the word "evaluating" for no benefit, since the repo has no `rustfmt.toml` and `format_strings` is off.

**Explicitly declined as out of scope:** extracting a shared helper across the two sibling timeout tests (would restructure a test this PR never touches), and tightening `cost_calls < 2` to `assert_eq!(cost_calls, 1)` (would reintroduce the flake that 547edce and e72a477 exist to fix — zero cost calls stays deliberately valid).

No behavior change; the assertion bound is untouched.

**Verification:** `cargo fmt -- --check` clean; failure message confirmed to render as `saw 0 cost calls` with a single space; `run_length_product_sets_stop_when_nested_deadline_expires` and `run_length_two_sets_stop_when_inner_deadline_expires` each 10/10 green (timing tests, so repeated rather than run once); `cargo test --lib search::enumerative` 42/42; `python3 -m unittest discover -s scripts -p "test_*_policy.py"` and `./scripts/test_test_all.sh` both pass.
